stardis/plasma/molecules.py

```diff
@@ -0,0 +1,422 @@
+    Multiplies the number density of each constituent ion and divides by the number density constant. Negative ions are ignored.
+    Attributes
+    ----------
+    molecule_number_density : DataFrame
+            A pandas DataFrame with dtype float. This represents the number density of each molecule at each depth point.
+    molecule_ion_map : DataFrame
+            A pandas DataFrame with the constituent ions of each molecule. Useful for calculating molecular masses later for doppler broadening.
+    """
+
+    # Need to think about negative ions - ignoring for now
+    # applicable for equilibrium constants given by Barklem and Collet 2016, which are given in SI units
+
+    outputs = ("molecule_number_density", "molecule_ion_map")
+
+    def calculate(self, ion_number_density, t_electrons, atomic_data):
+        # Preprocessing - split ions into symbol, charge, and number
+        try:
+            molecules_df = atomic_data.molecule_data.dissociation_energies.copy()
+        except:
+            raise ValueError(
+                "No molecular dissociation energies found in atomic data. Use Carsus to generate atomic data with the Barklem and Collet 2016 data."
+            )
+        molecules_df[
+            ["Ion1_symbol", "Ion1_positive", "Ion1_negative"]
+        ] = molecules_df.Ion1.str.extract(r"([A-Z][a-z]?+)(\+*)(\-*)")
+        molecules_df[
+            ["Ion2_symbol", "Ion2_positive", "Ion2_negative"]
+        ] = molecules_df.Ion2.str.extract(r"([A-Z][a-z]?+)(\+*)(\-*)")
+
+        molecules_df["Ion1"] = molecules_df["Ion1_symbol"].apply(
+            element_symbol2atomic_number
+        )
+        molecules_df["Ion2"] = molecules_df["Ion2_symbol"].apply(
+            element_symbol2atomic_number
+        )
+        molecules_df["Ion1_charge"] = molecules_df["Ion1_positive"].apply(
+            len
+        ) - molecules_df["Ion1_negative"].apply(len)
+        molecules_df["Ion2_charge"] = molecules_df["Ion2_positive"].apply(
+            len
+        ) - molecules_df["Ion2_negative"].apply(len)
+
+        number_densities_arr = np.zeros(
+            (len(atomic_data.molecule_data.equilibrium_constants), len(t_electrons))
+        )
+        # Get the temperatures at which the equilibrium constants are given for interpolation
+        equilibrium_const_temps = (
+            atomic_data.molecule_data.equilibrium_constants.columns.values
+        )
+        included_elements = ion_number_density.index.get_level_values(0).unique()
+
+        for (
+            molecule,
+            molecule_row,
+        ) in (
+            molecules_df.iterrows()
+        ):  # Loop over all molecules, calculate number densities using Barklem and Collet 2016 equilibrium constants - if a component ion does not exist in the plasma or is negative, assume no molecule
+            if (molecule_row.Ion1_charge == -1) or (molecule_row.Ion2_charge == -1):
```

Comment:
Could this be done with a filter of some kind instead of in a loop?

Reply:
Yes probably. Wolfgang has generally urged to make something that works before spending too much time thinking about how it can be optimized, but this almost certainly would be more efficient with a function that can take vector inputs, masking out the molecules we can't or don't want to handle, and then only operating on the useful ones and then mapping them back on to the original array. I couldn't quite figure out how to do that quickly but that would be worth doing if/when we think about optimizing this code.